server/modlog.ts

```diff
@@ -324,24 +466,103 @@ export class Modlog {
 		if (search.ip) regexString += `${this.escapeRegex(`[${search.ip}`)}.*?\\].*?`;
 		if (search.actionTaker) regexString += `${this.escapeRegex(`by ${search.actionTaker}`)}.*?`;
 		if (search.note) {
-			const regexGenerator = search.note.isExact ? this.generateRegex : this.escapeRegex;
+			const regexGenerator = search.note.isExact ? this.generateIDRegex : this.escapeRegex;
 			for (const noteSearch of search.note.searches) {
 				regexString += `${regexGenerator(toID(noteSearch))}.*?`;
 			}
 		}
 
-		const query = {
+		return {
 			rooms: rooms,
 			regexString,
 			maxLines: maxLines,
 			onlyPunishments: onlyPunishments,
 		};
-		const response = await PM.query(query);
+	}
 
-		if (response.duration > LONG_QUERY_DURATION) {
-			Monitor.log(`Long modlog query took ${response.duration} ms to complete: ${query}`);
+	prepareSQLSearch(
+		rooms: ModlogID[],
+		maxLines: number,
+		onlyPunishments: boolean,
+		search: ModlogSearch
+	): ModlogSQLQuery<string | number> {
+		for (const room of [...rooms]) {
+			rooms.push(`global-${room}` as ModlogID);
 		}
-		return {results: response, duration: response.duration};
+		const userid = toID(search.user);
+
+		const args: (string | number)[] = [];
+
+		let roomChecker = `roomid IN (${this.formatArray(rooms, args)})`;
+		if (rooms.includes('global')) roomChecker = `(roomid LIKE 'global-%' OR ${roomChecker})`;
+
+		let query = `SELECT *, (SELECT group_concat(userid, ',') FROM alts WHERE alts.modlog_id = modlog.modlog_id) as alts FROM modlog`;
+		query += ` WHERE ${roomChecker}`;
+
+		if (search.anyField) {
+			query += ` AND action LIKE '%' || ? || '%'`;
+
+			query += ` OR userid LIKE '%' || ? || '%'`;
+			query += ` OR autoconfirmed_userid LIKE '%' || ? || '%'`;
+			query += ` OR EXISTS(SELECT * FROM alts WHERE alts.modlog_id = modlog.modlog_id AND userid LIKE '%' || ? || '%')`;
+
+			query += ` OR ip LIKE '%' || ? || '%'`;
+			query += ` OR action_taker_userid LIKE '%' || ? || '%'`;
```

Comment:
Substring is always slow, if you want to do anything beyond a prefix match then you need to add an FTS column or manually tokenize.